### PR TITLE
lomiri.history-service: init at 0.4

### PR DESCRIPTION
**Pull request by OPNA2608**
## Description of changes

Working towards #99090.

[history-service](https://gitlab.com/ubports/development/core/hfd-service) provides call and conversation logs for the Lomiri stack. `telephony-service` and phone-related apps will depend on it.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/desktops/lomiri/services/history-service/default.nix [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+  ];
+
+  patches = [
+    # Deprecation warnings with Qt5.15, allow disabling -Werror
+    # Remove when version > 0.4
+    (fetchpatch {
+      url = "https://gitlab.com/ubports/development/core/history-service/-/commit/1370777952c6a2efb85f582ff8ba085c2c0e290a.patch";
+      hash = "sha256-Z/dFrFo7WoPZlKto6wNGeWdopsi8iBjmd5ycbqMKgxo=";
+    })
+    ./0001-Drop-deprecated-qt5_use_modules.patch
```
Comment (pbsds) [inline pkgs/desktops/lomiri/services/history-service/default.nix:52]: Can this be upstreamed and fetched?

**Review by pbsds — APPROVED** [review]
Result of `nixpkgs-review pr 261031` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>lomiri.history-service</li>
    <li>lomiri.history-service.dev</li>
  </ul>
</details>

LGTM, just a nit

File: pkgs/desktops/lomiri/services/history-service/default.nix (hunk 1 above)
Comment (OPNA2608) [reply pkgs/desktops/lomiri/services/history-service/default.nix]: Submitted & `fetchpatch`ed.

**Comment by pbsds** [comment]
Any idea what the [aarch64-linux](https://logs.ofborg.org/?key=nixos/nixpkgs.261031&attempt_id=f8361d69-fbd4-4f94-9511-73e0df965b4e) failure might be? Is it flaky?

**Comment by OPNA2608** [comment]
Seems to be flaky yeah. [That test](https://gitlab.com/ubports/development/core/history-service/-/blob/d31262e143cf307b91ad64ed89dc7bc306c8deb0/tests/daemon/DaemonTest.cpp#L586-593), along with afew other ones, uses QSignalSpy, which seem abit finnicky to use 100% reliably. [QSignalSpy tests also frequently fail on upstream CI](https://ci.ubports.com/job/UBportsCore/job/history-service/job/main/26/execution/node/48/log/) (grep for `FAIL!`), the solution there is to just wait for the next run...

Unfortunately, I would really prefer to have some tests enabled on this package. Figuring out the issue with the broken schema generation was really hard before I started running the test suite, and running it seems even more important now that we use a custom schema generation script. I'll try to check if we can disable them without losing the tests for the schemas.

**Comment by pbsds** [comment]
Your thorough work is exemplary. It is a joy reviewing these

**Comment by OPNA2608** [comment]
I have opened an issue about the flakiness upstream, since this also affects their infra: https://gitlab.com/ubports/development/core/history-service/-/issues/13

One possible, easy workaround would be to increase the amount of time all of those tests wait for the signal they're checking for for to be emitted. The current timeout is the default 5s one, Qt's test infra actually waits for 2x that and prints an additional message if the comparison passes after the intended timeout, but before 2x of it.

Unfortunately, building `history-service` on my current hardware takes at least 17 minutes, so this is kind of a chore to test. And in 4/4 local runs, I didn't hit this issue a single time. And the failure on OfBorg doesn't print this extra message, so it's hard to tell what happened or when the signal would've been emitted.

I'll try vendoring a patch to increase the per-check timeout to something higher - maybe 20 seconds for now, in case the test that timed out on OfBorg would've passed after the arbitrary 2x timeout extension. If that works locally & on OfBorg, then I'll test it afew more times when I'm back home. If it also works there, then I'd be in favour of keeping this vendored for awhile and watching Hydra to see if it ever has any issues like this. If that all works, then I'll send it upstream to see if they think it's good enough. Otherwise, I'll try to disable these tests and see if upstream has any opinions on how to solve this.

**Comment by OPNA2608** [comment]
Disabled the problematic tests, submitted some fixes upstream and cleaned up some other stuff. I think this should be good for review again.

**Comment by pbsds** [comment]
Result of `nixpkgs-review pr 261031` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>lomiri.history-service</li>
    <li>lomiri.history-service.dev</li>
  </ul>
</details>

Sorry I'm slow to respond during the holidays, as I'm mostly AFK. LGTM!

**Comment by OPNA2608** [comment]
No worries, I'm mostly busy with holidays & other projects myself.

**Pull request merged**